File: test-route-optimization.py
#!/usr/bin/env python3
"""Test script for the improved route optimization with weighted objective function."""

import sys
import os
sys.path.append('scheduler-api')

# Set environment variables
os.environ['SUPABASE_URL'] = 'https://ebyvourlaomcwitpibdl.supabase.co'
os.environ['SUPABASE_KEY'] = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImVieXZvdXJsYW9tY3dpdHBpYmRsIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NTAxODY4NDYsImV4cCI6MjA2NTc2Mjg0Nn0.jNK_J5mLw4DKZO8NTmXQzA5d51ugm9czzNkAEWyd4gY'
os.environ['GOOGLE_MAPS_API_KEY'] = 'AIzaSyA2MzeziWPYVyzwSLstnDySmqqm6oxz6FA'

from services.route_optimizer_service import RouteOptimizerService

def test_route_optimization():
    """Test improved route optimization with weighted objective function."""
    
    # Test data with different priorities
    customers = [
        {
            "id": "1",
            "name": "Anna Andersson", 
            "address": "Kungsgatan 45, Stockholm",
            "latitude": 59.3348413,
            "longitude": 18.0612716,
            "service_time": 30,
            "priority": "high"  # High priority
        },
        {
            "id": "2", 
            "name": "Bengt Bengtsson",
            "address": "Drottninggatan 89, Stockholm", 
            "latitude": 59.3293235,
            "longitude": 18.0685808,
            "service_time": 45,
            "priority": "low"   # Low priority
        },
        {
            "id": "3",
            "name": "Cecilia Carlsson",
            "address": "Sveavägen 24, Stockholm",
            "latitude": 59.3364022, 
            "longitude": 18.0624000,
            "service_time": 20,
            "priority": "high"  # High priority
        },
        {
            "id": "4",
            "name": "David Davidsson", 
            "address": "Götgatan 78, Stockholm",
            "latitude": 59.3142764,
            "longitude": 18.0746394,
            "service_time": 60,
            "priority": "medium"  # Medium priority
        }
    ]
    
    # Initialize route optimizer
    optimizer = RouteOptimizerService()
    
    print("🧪 Testing improved route optimization with weighted objective function")
    print("=" * 70)
    
    # Test 1: Distance optimization
    print("\n📍 Test 1: Distance optimization with priority weighting")
    print("Customers:")
    for customer in customers:
        print(f"  - {customer['name']}: {customer['priority']} priority ({customer['service_time']} min service)")
    
    try:
        result = optimizer.optimize_route(
            customers=customers,
            optimization_criteria="minimize_distance",
            depot_coordinates=(59.3293235, 18.0685808),  # Central Stockholm
            google_maps_api_key="AIzaSyA2MzeziWPYVyzwSLstnDySmqqm6oxz6FA"
        )
        
        print(f"\n✅ Distance optimization completed!")
        print(f"📊 Results:")
        print(f"   Total distance: {result['total_distance']:.2f} km")
        print(f"   Optimization time: {result['optimization_time']:.3f} seconds")
        print(f"   Objective value: {result['objective_value']:.2f}")
        
        print(f"\n🗺️ Optimized route order:")
        for i, customer in enumerate(result['route_order']):
            priority_icon = "🔴" if customer['priority'] == "high" else "🟡" if customer['priority'] == "medium" else "🟢"
            print(f"   {i+1}. {customer['name']} {priority_icon} ({customer['priority']} priority)")
            
    except Exception as e:
        print(f"❌ Distance optimization failed: {str(e)}")
    
    # Test 2: Time optimization  
    print("\n⏰ Test 2: Time optimization with priority weighting")
    
    try:
        result = optimizer.optimize_route(
            customers=customers,
            optimization_criteria="minimize_time", 
            depot_coordinates=(59.3293235, 18.0685808),
            vehicle_speed_kmh=35.0,  # Stockholm city speed
            google_maps_api_key="AIzaSyA2MzeziWPYVyzwSLstnDySmqqm6oxz6FA"
        )
        
        print(f"\n✅ Time optimization completed!")
        print(f"📊 Results:")
        print(f"   Total time: {result.get('total_time', 'N/A')} minutes")
        print(f"   Total distance: {result['total_distance']:.2f} km") 
        print(f"   Optimization time: {result['optimization_time']:.3f} seconds")
        print(f"   Objective value: {result['objective_value']:.2f}")
        
        print(f"\n🗺️ Optimized route order:")
        for i, customer in enumerate(result['route_order']):
            priority_icon = "🔴" if customer['priority'] == "high" else "🟡" if customer['priority'] == "medium" else "🟢"
            print(f"   {i+1}. {customer['name']} {priority_icon} ({customer['priority']} priority)")
            
    except Exception as e:
        print(f"❌ Time optimization failed: {str(e)}")

if __name__ == "__main__":
    test_route_optimization()

File: test-route-optimization.mjs
/**
 * Test route optimization backend integration
 * Tests the /api/route/optimize-route endpoint
 */

const API_URL = process.env.VITE_SCHEDULER_API_URL || 'https://mittschema-gurobi-backend.onrender.com';

async function testRouteOptimization() {
  console.log('🧪 Testing Route Optimization Backend\n');
  console.log(`Using API: ${API_URL}\n`);

  // Test 1: Health check
  console.log('📋 Test 1: Health Check');
  try {
    const healthResponse = await fetch(`${API_URL}/api/route/health`);
    const healthData = await healthResponse.json();
    console.log('✅ Health check:', JSON.stringify(healthData, null, 2));
  } catch (error) {
    console.error('❌ Health check failed:', error.message);
  }

  // Test 2: Demo customers
  console.log('\n📋 Test 2: Load Demo Customers');
  try {
    const demoResponse = await fetch(`${API_URL}/api/route/demo-customers`);
    const demoData = await demoResponse.json();
    console.log(`✅ Loaded ${demoData.customers?.length || 0} demo customers`);
    console.log('First customer:', JSON.stringify(demoData.customers?.[0], null, 2));
  } catch (error) {
    console.error('❌ Demo customers failed:', error.message);
  }

  // Test 3: Geocoding
  console.log('\n📋 Test 3: Geocode Address');
  try {
    const geocodeResponse = await fetch(`${API_URL}/api/route/geocode`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ address: 'Drottninggatan 1, Stockholm' })
    });
    const geocodeData = await geocodeResponse.json();
    console.log('✅ Geocoded address:', JSON.stringify(geocodeData, null, 2));
  } catch (error) {
    console.error('❌ Geocoding failed:', error.message);
  }

  // Test 4: Route Optimization
  console.log('\n📋 Test 4: Optimize Route');
  const testCustomers = [
    {
      id: '1',
      name: 'Anna Andersson',
      address: 'Vasagatan 10, Stockholm',
      latitude: 59.3326,
      longitude: 18.0649,
      serviceTime: 30,
      priority: 'high'
    },
    {
      id: '2',
      name: 'Bengt Bengtsson',
      address: 'Kungsgatan 20, Stockholm',
      latitude: 59.3345,
      longitude: 18.0632,
      serviceTime: 45,
      priority: 'medium'
    },
    {
      id: '3',
      name: 'Cecilia Carlsson',
      address: 'Sveavägen 30, Stockholm',
      latitude: 59.3363,
      longitude: 18.0586,
      serviceTime: 30,
      priority: 'medium'
    }
  ];

  try {
    const optimizeResponse = await fetch(`${API_URL}/api/route/optimize-route`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        customers: testCustomers,
        optimization_criteria: 'minimize_distance',
        startLocation: [59.3293, 18.0686], // Stockholm centrum
        max_route_time: 480,
        vehicle_speed_kmh: 40.0
      })
    });

    const optimizeData = await optimizeResponse.json();
    
    if (optimizeData.success) {
      console.log('✅ Route optimization successful!');
      console.log(`   Total Distance: ${optimizeData.totalDistance?.toFixed(2)} km`);
      console.log(`   Total Time: ${optimizeData.totalTime?.toFixed(0)} minutes`);
      console.log(`   Optimized Order:`);
      optimizeData.customers?.forEach((customer, index) => {
        console.log(`   ${index + 1}. ${customer.name} - ${customer.address}`);
      });
      
      if (optimizeData.routeInstructions) {
        console.log('\n   Route Instructions:');
        optimizeData.routeInstructions.slice(0, 3).forEach(instruction => {
          console.log(`   - ${instruction}`);
        });
      }
    } else {
      console.log('⚠️  Route optimization failed (using fallback)');
      console.log('   Error:', optimizeData.error);
      if (optimizeData.fallback_route) {
        console.log(`   Fallback Distance: ${optimizeData.fallback_route.totalDistance?.toFixed(2)} km`);
        console.log(`   Fallback Time: ${optimizeData.fallback_route.totalTime?.toFixed(0)} minutes`);
      }
    }
  } catch (error) {
    console.error('❌ Route optimization failed:', error.message);
  }

  console.log('\n🎉 Tests complete!\n');
}

// Run tests
testRouteOptimization().catch(console.error);
